bundles/org.openhab.binding.plugwiseha/src/main/java/org/openhab/binding/plugwiseha/internal/PlugwiseHAHandlerFactory.java

```diff
@@ -0,0 +1,161 @@
+        } else if (PlugwiseHAZoneHandler.supportsThingType(thingTypeUID)) {
+            return super.createThing(thingTypeUID, configuration, thingUID, bridgeUID);
+        } else if (PlugwiseHAApplianceHandler.supportsThingType(thingTypeUID)) {
+            return super.createThing(thingTypeUID, configuration, thingUID, bridgeUID);
+        }
+
+        throw new IllegalArgumentException(
+                "The thing type " + thingTypeUID + " is not supported by the plugwiseha binding.");
+    }
+
+    // Protected and private methods
+
+    /**
+     * Creates a {@link ThingHandler} for the given thing.
+     *
+     * @param thing the thing
+     * @return thing the created handler
+     */
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (PlugwiseHABridgeHandler.supportsThingType(thingTypeUID)) {
+            this.logger.debug("Creating new Plugwise Home Automation Bridge");
+            PlugwiseHABridgeHandler bridge = new PlugwiseHABridgeHandler((Bridge) thing, this.httpClient);
+            registerPlugwiseHADiscoveryService(bridge);
+            return bridge;
+        } else if (PlugwiseHAZoneHandler.supportsThingType(thingTypeUID)) {
+            logger.debug("Creating new Plugwise Home Automation Zone");
+            return new PlugwiseHAZoneHandler(thing);
+        } else if (PlugwiseHAApplianceHandler.supportsThingType(thingTypeUID)) {
+            logger.debug("Creating new Plugwise Home Automation Appliance");
+            return new PlugwiseHAApplianceHandler(thing);
+        }
+        return null;
+    }
+
+    private synchronized void registerPlugwiseHADiscoveryService(PlugwiseHABridgeHandler bridgeHandler) {
+        Hashtable<String, Object> properties = new Hashtable<String, Object>();
+        PlugwiseHADiscoveryService discoveryService = new PlugwiseHADiscoveryService(bridgeHandler);
+
+        ServiceRegistration<?> serviceRegistration = bundleContext.registerService(DiscoveryService.class.getName(),
+                discoveryService, properties);
+
+        this.discoveryServiceRegs.put(bridgeHandler.getThing().getUID(), serviceRegistration);
+
+        discoveryService.activate();
+    }
+
+    @Override
+    protected synchronized void removeHandler(ThingHandler thingHandler) {
+        if (thingHandler instanceof PlugwiseHABridgeHandler) {
+            ServiceRegistration<?> serviceReg = this.discoveryServiceRegs.remove(thingHandler.getThing().getUID());
+            if (serviceReg != null) {
+                PlugwiseHADiscoveryService service = (PlugwiseHADiscoveryService) bundleContext
+                        .getService(serviceReg.getReference());
+
+                if (service != null) {
+                    service.deactivate();
+                }
+                serviceReg.unregister();
+                discoveryServiceRegs.remove(thingHandler.getThing().getUID());
+            }
+        }
+    }
```

Comment:
This can all be removed by implementing the discovery as a `ThingHandlerService` See this preview documentation: https://deploy-preview-1262--openhab-docs-preview.netlify.app/docs/developer/bindings/#discovery-that-is-bound-to-a-thing

Reply:
If i do so i get this warning:
`Redundant superinterface DiscoveryService for the type PlugwiseHADiscoveryService, already defined by AbstractDiscoveryService`
I looked at some other bindings, but without success. Sorry, but could you help me in the good direction?

Reply:
I tried to use that documentation as well and had an error too.   I think that may need to be revisited from a documentation standpoint, what is there doesn't work without error in the latest code.

Reply:
Does discovery still works if you remove it? Because asaik it was/is required to make it work using the ThingHanderService.